### PR TITLE
also resolve conversations when dismissing reviews

### PR DESCRIPTION
**Pull request by renefritze**
Resolving conversations is not available through the Rest API so we need to use the GraphQL one. This is the first time using it for me.
The mutation query needs the token to have `contents: write`. I found no documentation for this. Basically had to bisect the range of all possible perms to find that.
I've made the missing permission not fail the action to allow a "soft" upgrade path. An annotation is added to the github actions run. Not sure if that's a good path, it's not super obvious where that message comes from. Could use a link back to actual README?

[Example with full permission](https://github.com/renefritze/tidy_comments_test/pull/14)

[Example with missing permission](https://github.com/renefritze/tidy_comments_test/pull/15)

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
> The mutation query needs the token to have `contents: write`. I found no documentation for this. Basically had to bisect the range of all possible perms to find that.

Since this functionality requires additional permissions, I would suggest making it optional, enabled via a special action config parameter, and disabled by default. Not everyone needs it (for example, I don't).

**Comment by renefritze** [comment]
Sure, I'll try to do that. When the action config is on, but the query detects permission's missing, I'd turn that into a an error output and step failure then, right?

I'll also address the pylint hints.

**Comment by renefritze** [comment]
I think I have addressed everything

**Comment by renefritze** [comment]
I did another set of test PRs with the updated action.
[No auto resolving](https://github.com/renefritze/tidy_comments_test/pull/16)
[Auto resolving](https://github.com/renefritze/tidy_comments_test/pull/17)

File: run_action.py [hunk 1]
```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
...
+            }
+          }
+        }
+      }
+    }
+    """ % (
+        repo_owner,
+        repo_name,
+        pr_number,
+    )
```
Comment (platisd) [inline run_action.py:541]: Nitpick: Can you please use an `F string`  for clarity since this is rather long query?

File: run_action.py [hunk 2]
```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
...
+                    # which we get through the Rest API
+                    and comment["author"]["login"] == "github-actions"
+                ):
+                    yield thread
+                    break
+    else:
+        print(
+            f"::error::getting unresolved conversation threads: {response.status_code}"
+        )
+        raise RuntimeError("Failed to get unresolved conversation threads.")
```
Comment (platisd) [inline run_action.py:571]: If the permission is missing, will we get a `403` or whatever here?
If yes, maybe add it to the printout as a hint to the user? :thinking:

File: run_action.py [hunk 3]
```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
...
+        else:
+            print("Conversation closed successfully.")
+            return
+    else:
+        print(f"::error::GraphQL request failed: {response.status_code}")
+    print(
+        "::error:: Failed to close conversation. See log for details and "
+        "https://github.com/platisd/clang-tidy-pr-comments/blob/master/README.md for help"
+    )
+    raise RuntimeError("Failed to close conversation.")
```
Comment (platisd) [inline run_action.py:615]: Can we please refactor this to decrease the "depth" of the function? E.g. it should look like this, with multiple return statements and no nested `if`s:

```python
if response.status_code != 200:
    print(f"::error::GraphQL request failed: {response.status_code}")
    # ...
    raise RuntimeError("Failed to close conversation.")
if "errors" not in response.json():
    print("Conversation closed successfully.")
    return
msg = response.json()["errors"][0]["message"]
# ....
```

File: run_action.py [hunk 4]
```diff
@@ -538,6 +669,14 @@ def main():
         required=True,
         help="Number of suggestions per comment",
     )
+    parser.add_argument(
+        "--auto-resolve-conversations",
+        type=str,
+        required=False,
+        # this matches the YAML schema for the input
+        default="false",
```
Comment (platisd) [inline run_action.py:677]: Considering that all other arguments are required for the Python script, even if they are not required by the YAML and have a default value there, what about the following in the name of consistency:

```suggestion
        required=True,
        help="If 'true', then close any discussions opened by the Action",
```

File: run_action.py [hunk 5]
```diff
@@ -495,6 +494,138 @@ def dismiss_change_requests(
...
+            pull_request_id=pull_request_id,
+            github_api_timeout=github_api_timeout,
+        )
+
+
+def conversation_threads_to_close(repo, pr_number, github_token, github_api_timeout):
+    """Generator of unresolved conversation threads to close
+
+    Uses the GitHub GraphQL API to get conversation threads for the given PR.
+    Then filters for unresolved threads and those that have been created by the action.
```
Comment (platisd) [inline run_action.py:510]: Question: Where do we select those created by the action?

**Review by platisd — CHANGES_REQUESTED** [review]
Great job! :+1: :100: :1st_place_medal:

File: run_action.py (hunk 2 above)
Comment (renefritze) [reply run_action.py:571]: You'd expect that, yeah. But no. The response is still 200 in that case and the missing auth is only mentioned in the error string.

File: run_action.py (hunk 5 above)
Comment (renefritze) [reply run_action.py:510]: Fair point. I guess we select open conversations started by the github-action user: https://github.com/platisd/clang-tidy-pr-comments/pull/76/files#diff-9981508c066d4bcb876bb89f7092ac196ce743485c2782d4142e6eb5e7c34457R561-R563
which might select conversations not actually started by the action. There's no message prefix to check for those though, right? Do you have another idea how to make the selection tighter?

File: run_action.py (hunk 3 above)
Comment (renefritze) [reply run_action.py:615]: With the current flow we don't have to repeat the runtime error + print for the error cases in the `response.status_code == 200` branch.
Do you want me to pull the print+raise into a local function, do the reorder and call that function where needed?

File: run_action.py (hunk 1 above)
Comment (renefritze) [reply run_action.py]: Done. With the other query too. Had opted to use this format to not need to use double curlies.

Comment (platisd) [reply run_action.py]: Oh jeez, looks ugly now. Is it OK to take my suggestion back? :laughing:

File: run_action.py (hunk 5 above)
Comment (platisd) [reply run_action.py:510]: Maybe it's a good idea to exploit that every message [starts](https://github.com/platisd/clang-tidy-pr-comments/blob/master/run_action.py#L196C22-L196C31) with `⚠️ .... ⚠️`?

File: run_action.py (hunk 1 above)
Comment (renefritze) [reply run_action.py]: Sure. I've made the change a sep commit. I'll just drop that and force-push :)

File: run_action.py (hunk 5 above)
Comment (renefritze) [reply run_action.py:510]: Ah, yes. Maybe pull up that string to either module level "constant" or pass it around like the warning prefix? Just so the two places cannot drift?

File: run_action.py (hunk 1 above)
Comment (renefritze) [reply run_action.py:541]: Done

File: run_action.py (hunk 3 above)
Comment (platisd) [reply run_action.py:612]: I am OK with both the local function and also with this small duplication. :+1:

File: run_action.py (hunk 5 above)
Comment (renefritze) [reply run_action.py:526]: done. Tried to be explicit with using the kw-arg version in existing code.

**Comment by renefritze** [comment]
Tried a new round of check PRs. Looks like I borked the comment-warning bit as it's no longer auto-closing in the test. Will revist later.

File: run_action.py (hunk 5 above)
Comment (platisd) [reply run_action.py:526]: Maybe it's enough with the `single_comment_prefix` for now, but what about a regex that checks that the first line starts and finishes with `:warning:`?

Comment (renefritze) [reply run_action.py:525]: Updated to use a regex

**Comment by renefritze** [comment]
New round of PRs that check new beahviour:

[PR that only got the the change request closed](https://github.com/renefritze/tidy_comments_test/pull/25)
[PR with right perms that got conversation closed](https://github.com/renefritze/tidy_comments_test/pull/26)
[PR with auto-close enabled, but perms missing](https://github.com/renefritze/tidy_comments_test/actions/runs/8419036886)

File: action_launcher.bash [hunk 6]
```diff
@@ -28,4 +28,5 @@ cd "$recreated_repo_dir"
   --repository "$GITHUB_REPOSITORY" \
   --repository-root "$recreated_repo_dir" \
   --request-changes "$INPUT_REQUEST_CHANGES" \
-  --suggestions-per-comment "$INPUT_SUGGESTIONS_PER_COMMENT"
+  --suggestions-per-comment "$INPUT_SUGGESTIONS_PER_COMMENT" \
+  --auto-resolve-conversations "$INPUT_AUTO_RESOLVE_CONVERSATIONS" \
```
Comment (oleg-derevenetz) [inline action_launcher.bash:32]: There's an extra backslash.

**Review by platisd — APPROVED** [review]

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]

**Pull request merged**